runtime/drivers/duckdb/connectors.go

```diff
@@ -71,20 +123,82 @@ func (c *connection) ingest(ctx context.Context, env *connectors.Env, source *co
 	return summary, nil
 }
 
-// for files downloaded locally from remote sources
-func (c *connection) ingestIteratorFiles(ctx context.Context, source *connectors.Source, filenames []string, appendToTable bool) error {
-	from, err := sourceReader(filenames, source.Properties)
+// updateSchema updates the schema of the table in case new file adds a new column or
+// updates the datatypes of an existing columns with a wider datatype.
+func (c *connection) updateSchema(ctx context.Context, from string, fileNames []string, source *connectors.Source, oldSchema map[string]string,
+	allowAddition, allowRelaxation bool,
+) (srcSchema, currentSchema map[string]string, err error) {
+	// schema of new files
+	if srcSchema, err = c.scanSchemaFromQuery(ctx, fmt.Sprintf("DESCRIBE (SELECT * FROM %s LIMIT 0);", from)); err != nil {
+		return
+	}
+
+	// combined schema
+	qry := fmt.Sprintf("DESCRIBE ((SELECT * FROM %s limit 0) UNION ALL BY NAME (SELECT * FROM %s limit 0));", source.Name, from)
+	unionSchema, err := c.scanSchemaFromQuery(ctx, qry)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	var query string
-	if appendToTable {
-		query = fmt.Sprintf("INSERT INTO %q (SELECT * FROM %s);", source.Name, from)
-	} else {
-		query = fmt.Sprintf("CREATE OR REPLACE TABLE %s AS (SELECT * FROM %s);", source.Name, from)
+	// current schema
+	currentSchema = oldSchema
+	if currentSchema == nil {
+		currentSchema, err = c.scanSchemaFromQuery(ctx, fmt.Sprintf("DESCRIBE %s;", source.Name))
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	newCols := make(map[string]string)
+	colTypeChanged := make(map[string]string)
+	for colName, colType := range unionSchema {
+		oldType, ok := currentSchema[colName]
+		if !ok {
+			newCols[colName] = colType
+		} else if oldType != colType {
+			colTypeChanged[colName] = colType
+		}
+	}
+
+	if !allowRelaxation {
+		if len(srcSchema) < len(unionSchema) {
+			c.logger.Error("new files are missing columns and column relaxation not allowed",
+				zap.String("files", strings.Join(names(fileNames), ",")),
+				zap.String("columns", strings.Join(missingMapKeys(unionSchema, srcSchema), ",")))
+			return nil, nil, errors.New("new files are missing columns and schema relaxation not allowed")
+		}
+
+		if len(colTypeChanged) != 0 {
+			c.logger.Error("new files change datatypes of some columns and column relaxation not allowed",
+				zap.String("files", strings.Join(names(fileNames), ",")),
+				zap.String("columns", strings.Join(keys(colTypeChanged), ",")))
+			return nil, nil, errors.New("new files change datatypes of some columns and column relaxation not allowed")
+		}
+	}
+
+	if len(newCols) != 0 && !allowAddition {
+		c.logger.Error("new files have new columns and column addition not allowed",
+			zap.String("files", strings.Join(names(fileNames), ",")),
+			zap.String("columns", strings.Join(keys(newCols), ",")))
+		return nil, nil, errors.New("new files have new columns and column addition not allowed")
+	}
```

Comment:
Can we make the errors a little richer? Imagine if a user sees this error, it should help them resolve the problem. Maybe show the first filename and all the column names?

Reply:
I am adding that information in the log entry so a user should be able to get this information. These logs should be available on the UI as well right ?
Seems weird to put all that information in the error message.

Reply:
For cloud deployments people don't have access to the log messages.

So if a source breaks one day, and a dashboard stops refreshing, the errors shown in the UI will not be informative